src/lib.rs

```diff
@@ -293,6 +293,77 @@ pub fn resourceprep(s: &str) -> Result<Cow<'_, str>, Error> {
     Ok(Cow::Owned(normalized))
 }
 
+fn x520_mapped_to_nothing(c: char) -> bool {
+    if c.is_control() {
+        return true;
+    }
+    match c {
+        '\u{00AD}' | '\u{1806}' | '\u{034F}' | '\u{180B}'..='\u{180D}' |
+        '\u{FE00}'..='\u{FE0F}' | '\u{FFFC}' | '\u{200B}' => true,
+        _ => false,
+    }
+}
+
+/// Prepares a string according to the procedures described in Section 7 of
+/// [ITU-T Recommendation X.520 (2019)](https://www.itu.int/rec/T-REC-X.520-201910-I/en).
+///
+/// Note that this function does _not_ remove leading, trailing, or inner
+/// spaces as described in Section 7.6, because the characters needing removal
+/// will vary across the matching rules and ASN.1 syntaxes used.
+pub fn x520prep(s: &str, case_fold: bool) -> Result<Cow<'_, str>, Error> {
+    if s.chars().all(|c| matches!(c, ' '..='~') && (!case_fold || c.is_ascii_lowercase())) {
```

Comment:
The fast path also needs to exclude empty strings.

Reply:
See comment above about empty string handling.